src/lib/storage/fsst_segment/fsst_encoder.hpp

```diff
@@ -0,0 +1,192 @@
+#pragma once
+
+#include <algorithm>
+#include <array>
+#include <limits>
+#include <memory>
+#include <string>
+
+#include "storage/base_segment_encoder.hpp"
+#include "storage/value_segment.hpp"
+#include "storage/value_segment/value_segment_iterable.hpp"
+#include "types.hpp"
+#include "utils/assert.hpp"
+#include "utils/enum_constant.hpp"
+
+namespace opossum {
+
+class FSSTEncoder : public SegmentEncoder<FSSTEncoder> {
+ public:
+  static constexpr auto _encoding_type = enum_c<EncodingType, EncodingType::FSST>;
+  static constexpr auto _uses_vector_compression = true;
+
+  template <typename T>
+  bool _collect_values(const AnySegmentIterable<T>& segment_iterable, pmr_vector<T>& values,
+                       pmr_vector<bool>& null_values) {
+    bool has_null_values = false;
+
+    segment_iterable.with_iterators([&](auto it, auto end) {
+      // Early out for empty segments, code below assumes it to be non-empty
+      if (it == end) {
+        return;
+      }
+
+      const auto segment_size = static_cast<size_t>(std::distance(it, end));
+      values.reserve(segment_size);
+      null_values.reserve(segment_size);
+
+      T current_value;
+      bool is_current_null;
+
+      for (; it != end; ++it) {
+        auto segment_value = *it;
+
+        current_value = segment_value.value();
+        is_current_null = segment_value.is_null();
+        if (is_current_null) {
+          values.emplace_back("");
+          has_null_values = true;
+        } else {
+          values.emplace_back(current_value);
+        }
+        null_values.emplace_back(is_current_null);
+      }
+    });
+
+    // The resize method of the vector might have overallocated memory - hand that memory back to the system.
+    values.shrink_to_fit();
+    null_values.shrink_to_fit();
+
+    return has_null_values;
+  }
+
+  template <typename T>
+  fsst_decoder_t _compress_values(pmr_vector<T>& values, pmr_vector<unsigned char>& compressed_values,
+                                  pmr_vector<uint64_t>& compressed_value_lengths) {
+    // Temporary data structures keeping char pointers and their lengths.
+    std::vector<uint64_t> row_lengths;
+    std::vector<unsigned char*> row_pointers;
+    row_lengths.reserve(values.size());
+    row_pointers.reserve(values.size());
+
+    // Needed for compression.
+    pmr_vector<unsigned char*> compressed_value_pointers;
+    compressed_value_pointers.resize(values.size());
+    compressed_value_lengths.resize(values.size());
+
+    uint64_t total_length = 0LL;
```

Comment:
You could write `auto total_length = 0ul;`

Reply:
Or even more explicit: `auto total_length = uint64_t{0};